include/seastar/core/abort_source.hh

```diff
@@ -104,10 +104,9 @@ private:
     std::exception_ptr _ex;
 
     void do_request_abort(std::optional<std::exception_ptr> ex) noexcept {
-        assert(_subscriptions);
-        _ex = ex.value_or(get_default_exception());
+        _ex = _ex ? _ex : ex.value_or(get_default_exception());
         auto subs = std::exchange(_subscriptions, std::nullopt);
-        while (!subs->empty()) {
+        while (subs && !subs->empty()) {
```

Comment:
with the above suggestion this shouldn't be needed.

also, once we rely on `_ex` (like we eo for abort_requested), `_subscriptions` can change from optional to boost::intrusive::list and just walked here, with exchange